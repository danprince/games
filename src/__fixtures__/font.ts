import { Font } from "..";
import url from "./font.png";

export let font: Font = {
  url,
  glyphWidth: 5,
  glyphHeight: 6,
  lineHeight: 7,
  precoloredGlyphs: 8,
  glyphWidthsTable: {},
};
