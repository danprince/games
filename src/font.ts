import defaultFontUrl from "./font.png";

/**
 * A bitmap font in the format that the engine understands.
 */
export interface Font {
  /**
   * The url of the font's image.
   */
  url: string;
  /**
   * The default width for each glyph, in pixels.
   */
  glyphWidth: number;
  /**
   * The default height for each glyph, in pixels.
   */
  glyphHeight: number;
  /**
   * The height for each line, in pixels.
   */
  lineHeight: number;
  /**
   * Widths for glyphs with variable widths.
   */
  glyphWidthsTable: Record<string, number>;
}

/**
 * The font that games use by default.
 */
export let defaultFont: Font = {
  url: defaultFontUrl,
  glyphWidth: 5,
  glyphHeight: 6,
  lineHeight: 7,
  glyphWidthsTable: {
    m: 6,
    M: 6,
    W: 6,
    T: 6,
    V: 6,
    w: 6,
    I: 4,
    "1": 4,
    f: 4,
    l: 3,
    j: 3,
    "<": 4,
    ">": 4,
    "-": 4,
    "=": 4,
    "*": 4,
    "+": 4,
    "?": 4,
    "{": 4,
    "}": 4,
    '"': 4,
    "[": 3,
    "]": 3,
    "(": 3,
    ")": 3,
    "|": 3,
    "'": 3,
    "`": 3,
    i: 2,
    ":": 2,
    ",": 3,
    ".": 2,
    "!": 2,
  },
};
