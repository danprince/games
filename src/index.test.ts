import { afterEach, beforeEach, expect, test, vi } from "vitest";
import { join } from "node:path";
import { readFileSync } from "node:fs";
import { bounds, canvas, end, global, local, measure, start, view, _update, _reset, down, pressed, released, Buttons, pointer, tween, delta, delay, write, draw, draw9Slice, font, preload, restore, save, writeLine, fillRect, color, strokeRect, line, stamp, drawFlipped, cancelTweens } from "../src/index";
import { font2 } from "./__fixtures__/font2";
import { font as testFont } from "./__fixtures__/font";
import * as sprites from "./__fixtures__/sprites";

// JSDOM doesn't support PointerEvent, so use MouseEvent when dispatching in
// tests.
let PointerEvent = MouseEvent;

/**
 * Advance to the next frame with the given time delta.
 * @param dt Milliseconds since last frame.
 */
async function frame(dt: number = 15) {
  _update(dt);
}

/**
 * Images need to be converted to base64 so that JSDOM can load them.
 */
function readAsDataUrl(path: string): string {
  let relative = join(__dirname, path);
  let binary = readFileSync(relative, { encoding: "base64" });
  return `data:image/png;base64,${binary}`;
}

vi.mock("./font.png", () => ({
  default: readAsDataUrl("./font.png"),
}));

vi.mock("./__fixtures__/sprites.png", () => ({
  default: readAsDataUrl("./__fixtures__/sprites.png")
}));

vi.mock("./__fixtures__/font.png", () => ({
  default: readAsDataUrl("./__fixtures__/font.png"),
}));

vi.mock("./__fixtures__/font2.png", () => ({
  default: readAsDataUrl("./__fixtures__/font2.png"),
}));

afterEach(() => _reset());

beforeEach(() => {
  // Don't allow tests to edit the default drawing state (it can cause
  // subsequent tests to fail).
  save();
});

test("delta", async () => {
  await start();
  frame(20);
  expect(delta()).toBe(20);
});

test("start sets canvas size", async () => {
  await start({ width: 110, height: 120 });
  expect(canvas.width).toBe(110);
  expect(canvas.height).toBe(120);
});

test("start appends canvas to body", async () => {
  await start({ width: 110, height: 120 });
  expect(document.body.children[0]).toBe(canvas);
});

test("automatic canvas scale", async () => {
  window.innerWidth = 1000;
  window.innerHeight = 1000;
  await start({ width: 50, height: 100 });
  expect(canvas.style.width).toBe("500px");
  expect(canvas.style.height).toBe("1000px");
});

test("max canvas scale", async () => {
  window.innerWidth = 1000;
  window.innerHeight = 1000;
  await start({ width: 50, height: 100, maxCanvasScale: 3 });
  expect(canvas.style.width).toBe("150px");
  expect(canvas.style.height).toBe("300px");
});

test("measure text", () => {
  expect([
    "Hello world!",
    "!!!",
    "WWW",
    "Text with a \nnewline",
  ].map(text => measure(text))).toMatchSnapshot();
});

test("measure text with alternate font", () => {
  expect([
    "Hello world!",
    "!!!",
    "WWW",
    "Text with a \nnewline",
  ].map(text => measure(text, font2))).toMatchSnapshot();
});

test("writing text", async () => {
  await start({ width: 100, height: 100 });
  write("hello world", 0, 10, "red", "black");
  write("hello world", 30, 40, "green", "black");
  write("hello world", 60, 70, "blue", "black");
  expect(canvas).toMatchSnapshot();
});

test("writing text with cursor state", async () => {
  await start({ width: 100, height: 100 });
  write("p1", 0, 10, "red");
  write("and p2");
  write("and p3");
  expect(canvas).toMatchSnapshot();
});

test("writing text in lines", async () => {
  await start({ width: 100, height: 100 });
  writeLine("this is line 1", 0, 10);
  writeLine("and this is line 2");
  writeLine("finally, line 3");
  expect(canvas).toMatchSnapshot();
});

test("writing with alternate fonts", async () => {
  preload(font2);
  await start();
  write("This is the default font", 10, 10);

  save();
  font(font2);
  write("This is an alternate font", 10, 20);
  restore();

  write("Back to the default font", 10, 30);

  expect(canvas).toMatchSnapshot();
});

test("font with shaded glyphs", async () => {
  preload(testFont);
  await start({ width: 100, height: 100 });
  font(testFont);
  write("abc", 0, 0, "red");
  expect(canvas).toMatchSnapshot();
});

test("font with precolored glyphs", async () => {
  preload(testFont);
  await start({ width: 100, height: 100 });
  font(testFont);
  writeLine("\x01\x02\x03\x04\x05\x06");
  expect(canvas).toMatchSnapshot();
});

test("fillRect", () => {
  start({ width: 100, height: 100 });
  fillRect(0, 10, 20, 30, "red");
  color("blue");
  fillRect(50, 10, 20, 30);
  expect(canvas).toMatchSnapshot();
});

test("strokeRect", () => {
  start({ width: 100, height: 100 });
  strokeRect(0, 10, 20, 30, "red");
  color("blue");
  strokeRect(50, 10, 20, 30);
  expect(canvas).toMatchSnapshot();
});

test("line", () => {
  start({ width: 100, height: 100 });
  line(0, 10, 20, 30, "red");
  color("blue");
  line(50, 10, 20, 30);
  expect(canvas).toMatchSnapshot();
});

test("stamp", () => {
  start({ width: 100, height: 100 });
  stamp(0xEAFFF1, 0, 10, "red");
  color("blue");
  stamp(0xEAFFF1, 10, 10);
  expect(canvas).toMatchSnapshot();
});

test("drawing sprites", async () => {
  preload(sprites);
  await start({ width: 100, height: 100 });
  draw(sprites.green_man, 0, 0);
  draw(sprites.red_man, 10, 20);
  expect(canvas).toMatchSnapshot();
});

test("drawing flipped sprites", async () => {
  preload(sprites);
  await start({ width: 100, height: 100 });
  draw(sprites.green_man, 0, 0);
  drawFlipped(sprites.green_man, 20, 0, true, false);
  drawFlipped(sprites.green_man, 40, 0, false, true);
  drawFlipped(sprites.green_man, 60, 0, true, true);
  expect(canvas).toMatchSnapshot();
});

test("drawing nine slice sprites", async () => {
  preload(sprites);
  await start({ width: 100, height: 100 });
  draw9Slice(sprites.nine_slice, 0, 0, 10, 10);
  draw9Slice(sprites.nine_slice, 20, 20, 30, 40);
  draw9Slice(sprites.nine_slice, 60, 20, 5, 5);
  expect(canvas).toMatchSnapshot();
});

test("views", async () => {
  await start({ width: 100, height: 100 });
  view(10, 10, 50, 50);
  write("hello", 0, 0);
  end();
  expect(canvas).toMatchSnapshot();
});

test("nesting views", async () => {
  await start({ width: 100, height: 100 });
  view(20, 20);
  write("view 1", 0, 0);
  view(20, 20);
  write("view 2", 0, 0);
  end();
  end();
  expect(canvas).toMatchSnapshot();
});

test("global coordinates", () => {
  expect(global(0, 0)).toEqual({ x: 0, y: 0 });

  view(10, 20);
  expect(global(0, 0)).toEqual({ x: 10, y: 20 });
  expect(global(5, 5)).toEqual({ x: 15, y: 25 });
  end();

  expect(global(0, 0)).toEqual({ x: 0, y: 0 });
});

test("local coordinates", () => {
  expect(local(0, 0)).toEqual({ x: 0, y: 0 });

  view(10, 20);
  expect(local(10, 20)).toEqual({ x: 0, y: 0 });
  expect(local(15, 25)).toEqual({ x: 5, y: 5 });
  expect(local(5, 15)).toEqual({ x: -5, y: -5 });
  end();

  expect(global(0, 0)).toEqual({ x: 0, y: 0 });
});

test("bounds", () => {
  expect(bounds()).toEqual({ x: 0, y: 0, w: Infinity, h: Infinity });
  view(10, 20, 30, 40);
  expect(bounds()).toEqual({ x: 10, y: 20, w: 30, h: 40 });
  end();
  expect(bounds()).toEqual({ x: 0, y: 0, w: Infinity, h: Infinity });
});

test("keyboard button states", async () => {
  await start();

  expect(down("Enter")).toBe(false);
  expect(pressed("Enter")).toBe(false);
  expect(released("Enter")).toBe(false);

  window.dispatchEvent(new KeyboardEvent("keydown", { key: "Enter" }));
  expect(down("Enter")).toBe(true);
  expect(pressed("Enter")).toBe(true);

  frame();
  expect(down("Enter")).toBe(true);
  expect(pressed("Enter")).toBe(false);

  window.dispatchEvent(new KeyboardEvent("keyup", { key: "Enter" }));
  expect(down("Enter")).toBe(false);
  expect(released("Enter")).toBe(true);

  frame();
  expect(released("Enter")).toBe(false);
});

test("mouse button states", async () => {
  await start();

  expect(down(Buttons.MouseLeft)).toBe(false);
  expect(pressed(Buttons.MouseLeft)).toBe(false);
  expect(released(Buttons.MouseLeft)).toBe(false);

  window.dispatchEvent(new PointerEvent("pointerdown"));
  expect(down(Buttons.MouseLeft)).toBe(true);
  expect(pressed(Buttons.MouseLeft)).toBe(true);

  frame();
  expect(down(Buttons.MouseLeft)).toBe(true);
  expect(pressed(Buttons.MouseLeft)).toBe(false);

  window.dispatchEvent(new PointerEvent("pointerup"));
  expect(down(Buttons.MouseLeft)).toBe(false);
  expect(released(Buttons.MouseLeft)).toBe(true);

  frame();
  expect(released(Buttons.MouseLeft)).toBe(false);
});

test("pointer coordinates", async () => {
  await start({ width: 20, height: 40 });

  canvas.getBoundingClientRect = () =>
    ({ x: 20, y: 40, width: 200, height: 400 } as DOMRect);

  window.dispatchEvent(new PointerEvent("pointermove", {
    bubbles: true,
    clientX: 100,
    clientY: 200,
  }));

  expect(pointer()).toEqual({ x: 8, y: 16 });
});

test("tweens", async () => {
  await start();
  let object = { a: 0 };
  let promise = tween(object, { a: 100 }, { duration: 1000 });
  frame(100);
  expect(object.a).toBe(10);
  frame(500);
  expect(object.a).toBe(60);
  frame(500);
  expect(object.a).toBe(100);
  expect(promise).resolves.toBe(undefined);
});

test("tween with custom easing", async () => {
  await start();
  let object = { a: 0 };
  let promise = tween(
    object,
    { a: 100 },
    { duration: 1000, easing: t => (t < 0.5 ? 0 : 1) },
  );
  frame(100);
  expect(object.a).toBe(0);
  frame(500);
  expect(object.a).toBe(100);
  expect(promise).resolves.toBe(undefined);
});

test("tween with step function", async () => {
  await start();
  let object = { a: 0 };
  let step = vi.fn();
  tween(
    object,
    { a: 100 },
    { duration: 1000, step },
  );
  frame(100);
  expect(step).toHaveBeenCalledWith(0.1);
  frame(100);
  expect(step).toHaveBeenCalledWith(0.2);
});

test("cancel tweens", async () => {
  await start();

  let object = { a: 0, b: 0, c: 0 };
  let step = vi.fn();

  tween(
    object,
    { a: 100 },
    { duration: 100, id: "t", step },
  );

  tween(
    object,
    { b: 100 },
    { duration: 100, id: "t" },
  );

  tween(
    object,
    { c: 100 },
    { duration: 100 },
  );

  frame(50);
  expect(object).toEqual({ a: 50, b: 50, c: 50 });
  cancelTweens("t");
  expect(object).toEqual({ a: 100, b: 100, c: 50 });
  frame(50);
  expect(object).toEqual({ a: 100, b: 100, c: 100 });
  expect(step).toHaveBeenCalledTimes(2);
});

test("delay", async () => {
  let callback = vi.fn();
  delay(100).then(callback);
  await frame(50);
  expect(callback).not.toHaveBeenCalled();
  await frame(50);
  expect(callback).toHaveBeenCalled();
  await frame(50);
  expect(callback).toHaveBeenCalledTimes(1);
});
