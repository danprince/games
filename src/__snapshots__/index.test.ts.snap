// Vitest Snapshot v1

exports[`alternate fonts 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAUAAAAC0CAYAAADl5PURAAAABmJLR0QA/wD/AP+gvaeTAAAC4klEQVR4nO3d0W7rIAwA0PZq///LvU+VImTAJqTqlnOkaWqKDc0WiyrQPh4AAAAAAAAAAAAAAAD8Pa/g53H43bat5l4Zz9kcmbzR82dfXzZ+pa9Kfri150LMq4k7XmxRvtnzK+2jNr24V3CsmvP4uM3Txr2C9u2xaJzReHp9jXJVzzdQMJrNtLPC7Aym2r4yjnbGOss1im3zjF5771j0uDemlb6y+eH2fi7O35tJ7WpfGcOu2E/Oqszg4EJXF8Ar3v4en58Vykybtt1q/syxZ9AmW+QqMSv5AQAAAAAAgL+gcoewepd0x13Ib7+TeXbx9c4xrPZ1duH0t/+NoKv6TzvaiZDJt2OXwpmlNe+Y6rh7eY7HZ+ckKhSV81l5HdXdL5ndLaN+2uPwK/zblKe9IKLdFyuFa0eOUdvehVzNE+Xs5W93cWQKTK//bF8zUcxoEXjURvHj19lVAFuzHRSZC3NHjm/zfHxud0mvr15buJ2rCuBs32r1LedqjuwMb0ee7IcprIxl1P/uvt6iD2OI7DrHAAAAAAAAwKdlPll5FHtF21nsrjuTmXWC1b5Wx/qJvwHcQnUZzHttWbtEJVqw3LtQM0tJouKQXaycic0WkUzse7lIu0bx2G62OLk39qhdr69e35X8QMdsZ8boIqvMHkczo2wRHMW+HrnxZGOr52RUqGdW+8rmh1upfiT+6NvRrozdzfeEAOUCGO1ZnR1bKZTHXQ27vwvD94QoogAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAANzVfyM2DLQbOsKgAAAAAElFTkSuQmCC`;

exports[`drawing nine slice sprites 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAABiklEQVR4nO3cMUrDYBzG4X+kg2dwEOnipJMewAN4C8FDOHsIwVt4AA+gk04uIg6ewU0H26ISK2oSX5rnWVrKBw35kTbN1y9NzUy39l6qxf3DddP2Ov2YVL3FODi5ah9xuv8iynCaeYzn28fWAes7m3V5uu9IGcjaf28AHwkSRpAwk6qqu4vz2j48ah1wd3E+6AaN3WT+xI7P0FS9nfZu7B63Dni6OXOGNaDFjvbDEAAAAOCz3i6LfHUppmu/vbQz3dqLnJqefD/k55bO0XftF3P+i+0L/L9A50G+m6Pv2k937Psr27PHqCijnTFMnf8ZXZD7h+vm6easqjLnenr5Dkk3ixD1UTU3uiNkLjFG1YiDpBIkjCBhBAkjSBhBwggSRpAwgoQRJIwgYQQJI0gYQcIIEkaQMIKEESSMIGEECSNIGEHCCBJGkDCChBEkjCBhBAkjSJheliMsu0NdH++1SnpbH7JqO2oovayRWHaHuq4lroL6i9EuiwYAAAAAAAAAAAAAAAAAAIBV8ArspncfT6tI+wAAAABJRU5ErkJggg==`;

exports[`drawing nine slices 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAABiklEQVR4nO3cMUrDYBzG4X+kg2dwEOnipJMewAN4C8FDOHsIwVt4AA+gk04uIg6ewU0H26ISK2oSX5rnWVrKBw35kTbN1y9NzUy39l6qxf3DddP2Ov2YVL3FODi5ah9xuv8iynCaeYzn28fWAes7m3V5uu9IGcjaf28AHwkSRpAwk6qqu4vz2j48ah1wd3E+6AaN3WT+xI7P0FS9nfZu7B63Dni6OXOGNaDFjvbDEAAAAOCz3i6LfHUppmu/vbQz3dqLnJqefD/k55bO0XftF3P+i+0L/L9A50G+m6Pv2k937Psr27PHqCijnTFMnf8ZXZD7h+vm6easqjLnenr5Dkk3ixD1UTU3uiNkLjFG1YiDpBIkjCBhBAkjSBhBwggSRpAwgoQRJIwgYQQJI0gYQcIIEkaQMIKEESSMIGEECSNIGEHCCBJGkDCChBEkjCBhBAkjSJheliMsu0NdH++1SnpbH7JqO2oovayRWHaHuq4lroL6i9EuiwYAAAAAAAAAAAAAAAAAAIBV8ArspncfT6tI+wAAAABJRU5ErkJggg==`;

exports[`drawing sprites 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAABC0lEQVR4nO3ZsY2DMBQG4PiULRiAddJlgBTMlIIB6FgnAzAH153AsdAJJ9hRvq8zlfV+np8FpxOf6T5d5tJ7+Abn1MNl8btmDMdth5/SG2AtGYiuqMh9uszLIyte816rTtgqvK45xtORFRe+a8YgjOMY6pUx1CuzCqRrxmCAl5XshDgUHVOYa29FUsUXSCFbhRcKAAD8eVxvrscZkv/U/2tZ/HbofV55AZ/fK5MViK6oyON6m5dHVrxmn11v+FbhdU2e3UdWXPh26IMw8hnqlTHUK7MrkHbogwH+HllveByKjinMtbciqeILpJCtwgsFAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgJf4BYO6bb79zZGsAAAAAElFTkSuQmCC`;

exports[`fillRect 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAAkUlEQVR4nO3R0QmAQAwFwZz993xWIPhx4CIzBSQ8dgYAAADoWHtmHz04s07ee28f3TGzPtlxffGUZ4LECBIjSIwgMYLECBIjSIwgMYLECBIjSIwgMYLECBIjSIwgMYLECBIjSIwgMYLECBIjSIwgMYLECBIjSIwgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD8zg00zgQ8AVRzswAAAABJRU5ErkJggg==`;

exports[`line 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAAsklEQVR4nO3TSQ4DIQxFQe5/abKKwoJE3WnADFUnsPX0UwIAAAAmklPK0Te0kTf5I60eJeetYrytGWXDEKV1omy6ipr5oxwSojRnlINWUTNXlINDlOKjHL6KmrgoQnw1NopVXDImihC39ItiFX9rH0WIx9pEsYqmnkURoov7Uayiu+tRhBjmdxSrCFGPIkSoTxSrmIgQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAoBfBIGWb6sJwyQAAAABJRU5ErkJggg==`;

exports[`measure 1`] = `
[
  {
    "h": 7,
    "w": 52,
    "x": 0,
    "y": 0,
  },
  {
    "h": 7,
    "w": 6,
    "x": 0,
    "y": 0,
  },
  {
    "h": 7,
    "w": 18,
    "x": 0,
    "y": 0,
  },
  {
    "h": 14,
    "w": 59,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`measure text 1`] = `
[
  {
    "h": 7,
    "w": 52,
    "x": 0,
    "y": 0,
  },
  {
    "h": 7,
    "w": 6,
    "x": 0,
    "y": 0,
  },
  {
    "h": 7,
    "w": 18,
    "x": 0,
    "y": 0,
  },
  {
    "h": 14,
    "w": 59,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`measure text with alternate font 1`] = `
[
  {
    "h": 6,
    "w": 65,
    "x": 0,
    "y": 0,
  },
  {
    "h": 6,
    "w": 18,
    "x": 0,
    "y": 0,
  },
  {
    "h": 6,
    "w": 18,
    "x": 0,
    "y": 0,
  },
  {
    "h": 12,
    "w": 67,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`measure with alternate font 1`] = `
[
  {
    "h": 6,
    "w": 65,
    "x": 0,
    "y": 0,
  },
  {
    "h": 6,
    "w": 18,
    "x": 0,
    "y": 0,
  },
  {
    "h": 6,
    "w": 18,
    "x": 0,
    "y": 0,
  },
  {
    "h": 12,
    "w": 67,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`nested views 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAAqklEQVR4nO3ZQQ6DMAwEwLTi/19Or1UEFlWd4NKZCxEEg4zwZVsDAAAAAAAAAK7Ui9b6C/1g/X4u2hM1fLz3dp6LntOH47g+65HwLqVtk+pGzT5qag+u8aVoLO2Nnb0/KKp767EFAAAAAADQ8jN6+coHZmf0Z/YtsypTz5ad0ZeJj2dl6tlmZvRlPsYvWZHRlxhZAAAAAAAAAAAAAAAAAAAAAAAAAAAAXOoFBWU137Z0Se8AAAAASUVORK5CYII=`;

exports[`nesting views 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAAqklEQVR4nO3ZQQ6DMAwEwLTi/19Or1UEFlWd4NKZCxEEg4zwZVsDAAAAAAAAAK7Ui9b6C/1g/X4u2hM1fLz3dp6LntOH47g+65HwLqVtk+pGzT5qag+u8aVoLO2Nnb0/KKp767EFAAAAAADQ8jN6+coHZmf0Z/YtsypTz5ad0ZeJj2dl6tlmZvRlPsYvWZHRlxhZAAAAAAAAAAAAAAAAAAAAAAAAAAAAXOoFBWU137Z0Se8AAAAASUVORK5CYII=`;

exports[`stamp 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAAc0lEQVR4nO3UOQrAMAwEwJXJ/7/stHYgrpyjmGkWoUawoAQAAAD4kZ701Txs+npmhyNZlXClhKe1rw9g1pKkkhrzXtWc7FbJ/LLWpYwvSykAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACvOwEgqhAG1dFRpwAAAABJRU5ErkJggg==`;

exports[`strokeRect 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAAqUlEQVR4nO3awQ3CQAwAQR+i/5ZDBSh5XMhKzBRgy9qvZwAAAICOdcwc24fOrN0zzx3b75hZv79jd5A7Al/cvHnvHYHPvZ5YyneCxAgSI0iMIDGCxAgSI0iMIDGCxAgSI0iMIDGCxAgSI0iMIDGCxAgSI0iMIDGCxAgSI0iMIDGCxAgSI0jMe+bJB+ndnnmQBgAAAAAAAAAAAAAAAAAAAAAAAAAAAAD+3gczgBAyF5ts6AAAAABJRU5ErkJggg==`;

exports[`views 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAAd0lEQVR4nO3UMQrAIAwFUC29/5XTxUJpFYy4FN4DF8nPEImlAAAAAP8UC3WzGZojWR/lO+TeXSbPQ/ZBajv3UDPbEJ08L+eGHnVDDxaMtmH0jc3WAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMOsC26YV+N2WyAMAAAAASUVORK5CYII=`;

exports[`writing text 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAByklEQVR4nO3c0ZKDIAxA0dDph/vn2QcLBgwaq66tvWemq9AECqhPrCIAAAAAAOCDqIiOh3hcNOdIvT6v+C1neoqU0Wgaj+lVp6VQDr5IrDrt57KXv9am3mgRrIfIONok08LYwWp1mMsTmwKx0fy1Nu33d/Mw59X4kvkEhUIXHjFeflpp9HZrktpHRx5hnjHzyFCdytKem9jmtCj52gS0+eb7KszrU/zYr2WeCt3zXJaNsZ72zujNr7PG3f4j/QIAAAAAAAAAAAAAAAAA8M+G4K51GxfNOVKvz5N+y2M95ESDzAc2iG5arLVYr327wFv7n76/mWnQ0wQMzUeayfHP1yfQW5Be//26eJ87PI9u8A31Pt1hZ37PuMjR/LRyF8T6/CpLV1599flX7jy2N0H+HeDl+1e93+dJd8iVO8bZdQ8AAAAAAAAAAAAAAAAAwPE0uOPQxkVz9vuEvb0X0NefJFJ2Huprwm1dNz8Qq077uezlj3U/uiDVO7S1vvhzXU+ZWBu7YTuplz81ce3/h1wrzYub3nYfDOw+7tx33f/yghh5D7WairKvWv1ze6Uvtdm+7X45/xd3bn/TrnsAAAAAAAAAAAAAAAAAwE5/KBEtGwCQK3cAAAAASUVORK5CYII=`;

exports[`writing text in lines 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAABYElEQVR4nO3Y0Y7CIBAAQGr8/1/uPZxNKgFhG4jSzLxUcYsrFLo2JQAAAOD37I33tbZIn5H46Lmjcvi6Lb0nvb2Opbaz1udX4vfCZ7Xz9kJb7Tt78vs5n1bIXjj2XHlX40vfWeqvp9/lVsjzwjn5KmpdgdH4SA6383gdj62r90qODG5vfL59tuJ6V8dyqwQYLXozvVMZPXULrJW5rXIzb0+NmLuU0Vvh9XCRcrN0TqvPVn8rldF5H8NdKXtHW6mMnroyUvqfkGi52eMc2/oBM7arcw49Exz557/0EwC4h/MN7Zvl5ehSNXKz/hm1ZGs/Jm+vDeKVwa39X/mUR68lJuV4llUbvC293/T2U9tsrTxKcbV+pldHoxxl74xkI1XZ1f4jcUtMyqMd8qb0pDUvG0tX9tmnSZrxxHepFTJbPmhL7OMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACM8gen45y/YS1iBwAAAABJRU5ErkJggg==`;

exports[`writing text with cursor state 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAAwElEQVR4nO3VWw6DIBQFwEvj/rdMf2pr8BGkUmwy8wUELkciGgEAAAD8gRyR+5RtWnOHLDn6ZXmblgefItJr59pNl/PSRn9rXm2teWyrPSJLWeMoS7vl4e+195eu2rVjtfVq1v0yS/cb8uhZ/CL93sbzUqxv36WmiM9tKD9ZOSKn/cOoDXbmAeZ5qeiX4yOzHOX4Xqefd4u75Ii4VxYAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABjqCeHkReA6V7QgAAAAAElFTkSuQmCC`;

exports[`writing with alternate fonts 1`] = `data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAUAAAAC0CAYAAADl5PURAAAABmJLR0QA/wD/AP+gvaeTAAAC4klEQVR4nO3d0W7rIAwA0PZq///LvU+VImTAJqTqlnOkaWqKDc0WiyrQPh4AAAAAAAAAAAAAAAD8Pa/g53H43bat5l4Zz9kcmbzR82dfXzZ+pa9Kfri150LMq4k7XmxRvtnzK+2jNr24V3CsmvP4uM3Txr2C9u2xaJzReHp9jXJVzzdQMJrNtLPC7Aym2r4yjnbGOss1im3zjF5771j0uDemlb6y+eH2fi7O35tJ7WpfGcOu2E/Oqszg4EJXF8Ar3v4en58Vykybtt1q/syxZ9AmW+QqMSv5AQAAAAAAgL+gcoewepd0x13Ib7+TeXbx9c4xrPZ1duH0t/+NoKv6TzvaiZDJt2OXwpmlNe+Y6rh7eY7HZ+ckKhSV81l5HdXdL5ndLaN+2uPwK/zblKe9IKLdFyuFa0eOUdvehVzNE+Xs5W93cWQKTK//bF8zUcxoEXjURvHj19lVAFuzHRSZC3NHjm/zfHxud0mvr15buJ2rCuBs32r1LedqjuwMb0ee7IcprIxl1P/uvt6iD2OI7DrHAAAAAAAAwKdlPll5FHtF21nsrjuTmXWC1b5Wx/qJvwHcQnUZzHttWbtEJVqw3LtQM0tJouKQXaycic0WkUzse7lIu0bx2G62OLk39qhdr69e35X8QMdsZ8boIqvMHkczo2wRHMW+HrnxZGOr52RUqGdW+8rmh1upfiT+6NvRrozdzfeEAOUCGO1ZnR1bKZTHXQ27vwvD94QoogAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAANzVfyM2DLQbOsKgAAAAAElFTkSuQmCC`;
